### PR TITLE
Support custom registries

### PR DESCRIPTION
**Pull request by l4l**
Required for docs.rs with custom index location as stated in [this comment](https://github.com/rust-lang/docs.rs/issues/767#issuecomment-710296929).

UPD: Tested with docs-builder example for a private registry crate and it worked fine with `https://login:password@git.repo` url. Still doesn't work with ssh though.

### CODE REVIEW

File: src/crates/registry.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    fn index(&self) -> &str {
+        self.registry_index.as_str()
+    }
+
+    fn index_folder(&self) -> String {
+        // https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits
+        self.registry_index.as_str().replace(
+            &['/', '\\', '<', '>', '|', '?', '*', '"', ':', '.'][..],
+            &"-",
+        )
```
Comment (emilyalbini) [inline src/crates/registry.rs]: What we did in other parts of rustwide is using the `percent_encoding` crate to urlencode the paths. [See an example in the git repos support.](https://github.com/rust-lang/rustwide/blob/d2c0723c88a7b05715b4942226acebf54d0b6328/src/crates/git.rs#L53-L58) What I'd do is move the definition of `ENCODE_SET` out of that module and into `crate::utils`, and use the same mechanism here.

File: src/crates/registry.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        match self {
+            Registry::CratesIo => "cratesios-sources".into(),
+            Registry::Alternative(alt) => format!("{}-sources", alt.index()),
+        }
+    }
+
+    fn name(&self) -> String {
+        match self {
+            Registry::CratesIo => "crates.io".into(),
+            Registry::Alternative(_alt) => todo!(),
```
Comment (emilyalbini) [inline src/crates/registry.rs]: Can we replace this with the index URL?

File: src/crates/registry.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+pub enum Registry {
+    CratesIo,
+    Alternative(AlternativeRegistry),
+}
+
+impl Registry {
+    fn cache_folder(&self) -> String {
+        match self {
+            Registry::CratesIo => "cratesios-sources".into(),
+            Registry::Alternative(alt) => format!("{}-sources", alt.index()),
```
Comment (emilyalbini) [inline src/crates/registry.rs]: This should probably use the encoded path (from `alt.index_folder()`).

File: src/crates/registry.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+pub enum Registry {
+    CratesIo,
+    Alternative(AlternativeRegistry),
+}
+
+impl Registry {
+    fn cache_folder(&self) -> String {
+        match self {
+            Registry::CratesIo => "cratesios-sources".into(),
```
Comment (emilyalbini) [inline src/crates/registry.rs]: Let's maintain compatibility with the existing directories.

```suggestion
            Registry::CratesIo => "cratesio-sources".into(),
```

File: src/crates/registry.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+use flate2::read::GzDecoder;
+use log::info;
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Read};
+use std::path::{Path, PathBuf};
+use tar::Archive;
+
+static CRATES_ROOT: &str = "https://static.crates.io/crates";
+
+impl RegistryCrate {
```
Comment (emilyalbini) [inline src/crates/registry.rs:65]: Could you move the impl below the struct definition?

File: src/crates/mod.rs [hunk 6]
```diff
@@ -14,7 +14,7 @@ trait CrateTrait: std::fmt::Display {
 }
 
 enum CrateType {
-    CratesIO(cratesio::CratesIOCrate),
+    CratesIO(registry::RegistryCrate),
```
Comment (emilyalbini) [inline src/crates/mod.rs]: Maybe let's rename the varianto to `Registry`?

**Review by emilyalbini — CHANGES_REQUESTED** [review]
Thanks for the PR! I left some inline comments, and there are some more general changes to make:

* For all items that are not part of the public API, could you switch the visibility from `pub` to `pub(super)`? That way it's clear those are not available publicly, and we're guaranteed to get unused item warnings.
* It'd be nice to add a buildtest that fetches a crate from an alternative registry. One you can use is the [`foo` crate on `staging.crates.io`](https://staging.crates.io/crates/foo) ([index url](https://github.com/rust-lang/staging.crates.io-index)), crates.io's staging instance.
* Could you add a changelog item?

File: src/crates/mod.rs [hunk 7]
```diff
@@ -23,10 +23,21 @@ enum CrateType {
 pub struct Crate(CrateType);
 
 impl Crate {
+    /// Load a create from specified registry.
```
Comment (emilyalbini) [inline src/crates/mod.rs]:
```suggestion
    /// Load a crate from specified registry.
```

**Review by emilyalbini — COMMENTED** [review]
This looks better! Once a buildtest that fetches from staging.crates.io is added this is ready to land!

**Comment by l4l** [comment]
Oh, thanks for noticing, missed that commit at local rebase. Fortunately git remembers it

**Review by emilyalbini — APPROVED** [review]

**Comment by emilyalbini** [comment]
Thank you so much for this PR!

**Pull request merged**